src/libmongoc/src/mongoc/mongoc-scram.c

```diff
@@ -1150,4 +1229,176 @@ _mongoc_sasl_prep (const char *in_utf8, int in_utf8_len, bson_error_t *err)
+   // can be determined by the first byte
+   if ((*c & 0x80) == 0)
+      length = 1;
+   else if ((*c & 0xe0) == 0xc0)
+      length = 2;
+   else if ((*c & 0xf0) == 0xe0)
+      length = 3;
+   else if ((*c & 0xf8) == 0xf0)
+      length = 4;
+   else
+      length = 1;
+
+   return length;
+}
+
+int
+_mongoc_utf8_string_length (const char *s)
+{
+   const char *c = s;
+
+   int str_length = 0;
+   int utf8_char_length;
+
+   while (*c) {
+      utf8_char_length = _mongoc_utf8_char_length (c);
+
+      if (!_mongoc_utf8_is_valid (c, utf8_char_length))
+         return -1;
+
+      str_length++;
+      c += utf8_char_length;
+   }
+
+   return str_length;
+}
+
+
+bool
+_mongoc_utf8_is_valid (const char *c, int length)
+{
+   // Referenced table here:
+   // https://lemire.me/blog/2018/05/09/how-quickly-can-you-check-that-a-string-is-valid-unicode-utf-8/
+   switch (length) {
+   case 1:
+      return _mongoc_char_between_chars (*c, 0x00, 0x7F);
+   case 2:
+      return _mongoc_char_between_chars (*c, 0xC2, 0xDF) &&
+             _mongoc_char_between_chars (c[1], 0x80, 0xBF);
+   case 3:
+      // Four options, separated by ||
+      return (_mongoc_char_between_chars (*c, 0xE0, 0xE0) &&
+              _mongoc_char_between_chars (c[1], 0xA0, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xE1, 0xEC) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xED, 0xED) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0x9F) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xEE, 0xEF) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF));
+   case 4:
+      // Three options, separated by ||
+      return (_mongoc_char_between_chars (*c, 0xF0, 0xF0) &&
+              _mongoc_char_between_chars (c[1], 0x90, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[3], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xF1, 0xF3) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[3], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xF4, 0xF4) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0x8F) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[3], 0x80, 0xBF));
+   default:
+      return true;
+   }
+}
+
+
+bool
+_mongoc_char_between_chars (const char c, const char lower, const char upper)
+{
+   return (c >= lower && c <= upper);
+}
+
+bool
+_mongoc_is_code_in_table (unsigned int code,
```

Comment:
Suggest being specific and state `code_point` instead of simply `code` (ambiguous).